src/AutoMapper/MapperConfiguration.cs

```diff
@@ -239,6 +247,16 @@ private TypeMap GetTypeMap(TypePair initialTypes)
                 }
             }
 
+            if (!hasMapper.Value 
+                && Configuration.CreateMissingTypeMaps
+                && !(initialTypes.SourceType.IsAbstract() && initialTypes.SourceType.IsClass())
+                && !(initialTypes.DestinationType.IsAbstract() && initialTypes.DestinationType.IsClass())
+                && !ExcludedTypes.Contains(initialTypes.SourceType)
+                && !ExcludedTypes.Contains(initialTypes.DestinationType))
+            {
+                return Configuration.CreateInlineMap(_typeMapRegistry, initialTypes);

```

Comment:
I would lock(this) here, so all CreateMap calls are serialized, closed generics, conventions and inline.

Reply:
Sounds good. I got rid of the cache, it's just not obvious.

Reply:
Ah but that's the value factory for the locking concurrent dictionary. The lock is already there.

Reply:
Yes, I know there's a lock for the type pair, I'm just not sure that's enough. I'll have to think about it some more.

Reply:
I don't think it's enough. A GetType map for some other types might call CreateMap for these initialTypes. And then you have concurrent CreateMap calls again.

Reply:
Another thing, I don't think we need a Lazy here. A regular bool? should do. It's a local, only one thread will access it and the mappers collection is immutable.

Reply:
Ah you're right, I missed the other locks.

Reply:
I use the `hasMapper` in two places though? That would look a little odd, I might need a func or something.

Reply:
I don't have a problem with
     
     var hasMapper = (FindMapper(initialTypes) != null);

Or simply calling it twice as before.

Reply:
Done!